internal/db/txn.go

```diff
@@ -32,26 +37,179 @@ func (t *explicitTxn) Discard(ctx context.Context) {
 
 // transactionDB is a db that can create transactions.
 type transactionDB interface {
-	NewTxn(context.Context, bool) (datastore.Txn, error)
+	NewTxn(context.Context, bool) (client.Txn, error)
 }
 
-// ensureContextTxn ensures that the returned context has a transaction
-// and an identity.
+// ensureContextTxn ensures that the returned context has a transaction.
 //
 // If a transactions exists on the context it will be made explicit,
 // otherwise a new implicit transaction will be created.
 //
-// The returned context will contain the transaction and identity
+// The returned context will contain the transaction
 // along with the copied values from the input context.
 func ensureContextTxn(ctx context.Context, db transactionDB, readOnly bool) (context.Context, datastore.Txn, error) {
 	// explicit transaction
-	txn, ok := txnctx.TryGet(ctx)
+	ctxTxn, ok := datastore.CtxTryGetTxn(ctx)
 	if ok {
-		return InitContext(ctx, &explicitTxn{txn}), &explicitTxn{txn}, nil
+		switch txn := ctxTxn.(type) {
+		case *Txn:
+			return InitContext(ctx, &explicitTxn{txn}), &explicitTxn{txn}, nil
+		case *explicitTxn:
+			return InitContext(ctx, &explicitTxn{txn.Txn}), &explicitTxn{txn.Txn}, nil
+		}
 	}
-	txn, err := db.NewTxn(ctx, readOnly)
+	clientTxn, err := db.NewTxn(ctx, readOnly)
 	if err != nil {
-		return nil, txn, err
+		return nil, nil, err
 	}
+	txn := clientTxn.(*Txn) //nolint:forcetypeassert
 	return InitContext(ctx, txn), txn, nil
 }
+
+type Txn struct {
+	*datastore.BasicTxn
+	db *DB
```

Comment:
suggestion: Given that we are introducing this new type, it looks like we can simplify the code a little bit and convert `explicitTxn` to a bool on this object - atm the embedding of `*Txn` in `explicitTxn` seems a bit odd.  If for whatever reason the code remains the same, it might be worth expanding the documentation on either type to explain why they are needed.